astropy/cosmology/connect.py

```diff
@@ -0,0 +1,218 @@
+__doctest_skip__ = ["CosmologyRead", "CosmologyWrite"]
+
+
+class CosmologyRead(io_registry.UnifiedReadWrite):
+    """Read and parse data to a `~astropy.cosmology.Cosmology`.
+
+    This can *only* be called from `~astropy.cosmology.Cosmology`,
+    not any subclasses.
+
+    This function provides the Cosmology interface to the Astropy unified I/O
+    layer. This allows easily reading a file in supported data formats using
+    syntax such as::
+
+        >>> from astropy.cosmology import Cosmology
+        >>> cosmo1 = Cosmology.read('cosmo1.ecsv')
+        >>> cosmo2 = Cosmology.read('cosmo2.json')
+
+    Get help on the available readers using the ``help()`` method::
+
+      >>> Cosmology.read.help()  # Get help reading and list supported formats
+      >>> Cosmology.read.help('json')  # Get detailed help on JSON reader
+      >>> Cosmology.read.list_formats()  # Print list of available formats
+
+    See also: https://docs.astropy.org/en/stable/io/unified.html
+
+    Parameters
+    ----------
+    *args : (optional)
+        Positional arguments passed through to data reader. If supplied the
+        first argument is typically the input filename.
+    format : str (optional, keyword-only)
+        File format specifier.
+    **kwargs
+        Keyword arguments passed through to data reader.
+
+    Returns
+    -------
+    out : `~astropy.cosmology.Cosmology` subclass instance
+        `~astropy.cosmology.Cosmology` corresponding to file contents.
+
+    Methods
+    -------
+    from_mapping
+        Parse a mapping instance (e.g. dict) into a cosmology instance.
+    from_table
+        Parse a `~astropy.table.Table` instance into a cosmology instance.
+
+    Notes
+    -----
+    The cosmology class must be specified by name in the metadata (if a |Table|)
+    or as a field titled ``cosmology`` if a mapping.
+    Metadata can be included on the |Table| or as a key "meta" for a JSON.
+
+    Warns
+    -----
+    `~astropy.utils.exceptions.AstropyUserWarning`
+        If ``read`` is examined not from the Cosmology base class.
+    """
+
+    def __new__(cls, instance, cosmo_cls):
+        from astropy.cosmology.core import Cosmology
+
+        # warn that ``read`` can only be called from the base class.
+        # and return None to prevent usage.
+        if cosmo_cls is not Cosmology:
+            warnings.warn(("``Cosmology.read()`` must be called from the "
+                           "``Cosmology`` base class."),
+                           category=AstropyUserWarning)
+            return None  # cannot use .read on subclasses, only `Cosmology`.
+
+        return super().__new__(cls)
+
+    def __init__(self, instance, cosmo_cls):
+        super().__init__(instance, cosmo_cls, "read")
+
+    def __call__(self, *args, **kwargs):
+        cosmo = io_registry.read(self._cls, *args, **kwargs)
+        return cosmo
+
+    @staticmethod
+    def from_mapping(mapping, *, move_to_meta=False):
```

Comment:
We had this conversation already... (look above for reference). I already accepted that it was OK to put them as methods on read/write.

Reply:
Sorry that I missed that. I still think it is good to reconsider, especially now that `SkyCoord.to_table` and `SkyCoord.from_table` exist, while as far as I know there is no precedent in python for putting anything on what otherwise look like methods like `.read` and `.write` (well, we have `.help`, which I am very happy with, but it is not exactly discoverable without reading docs).

I think the bottom line is a question of what `Cosmology` really is; arguably it is closer to a `Model` than a `SkyCoord`, but it is not like `Model` has a `.read` method either.

Anyway, I'm far from trying to stop this - after all, I don't use `cosmology` at all - I just think the design and implementation here are surprising and complicated.